From b3f403b241e8aeeb171a707ec29fb1c4ee73ed0a Mon Sep 17 00:00:00 2001
From: John Jacques <john.jacques@intel.com>
Date: Fri, 1 Jul 2016 15:37:40 +0300
Subject: [PATCH 01/13] axxia: PCI Updates for 5600 Hardware

Signed-off-by: John Jacques <john.jacques@intel.com>
Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
[Original patch is from linux-yocto-4.1 axxia/base branch.]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
[Modification applied for WRLinux context.]
---
 arch/arm64/boot/dts/intel/axm5616-victoria.dts |    5 +++
 drivers/pci/host/pcie-axxia.c                  |   40 ++++++++++++++---------
 2 files changed, 29 insertions(+), 16 deletions(-)

diff --git a/arch/arm64/boot/dts/intel/axm5616-victoria.dts b/arch/arm64/boot/dts/intel/axm5616-victoria.dts
index 9a749ab..6fc458e 100644
--- a/arch/arm64/boot/dts/intel/axm5616-victoria.dts
+++ b/arch/arm64/boot/dts/intel/axm5616-victoria.dts
@@ -120,3 +120,8 @@
 		};
 	};
 };
+
+&pci0 {
+
+	status = "okay";
+};
diff --git a/drivers/pci/host/pcie-axxia.c b/drivers/pci/host/pcie-axxia.c
index d1d68dd..f309195 100644
--- a/drivers/pci/host/pcie-axxia.c
+++ b/drivers/pci/host/pcie-axxia.c
@@ -100,7 +100,6 @@
 
 /* SYSCON */
 #define AXXIA_SYSCON_BASE             0x8002C00000
-static void __iomem *axxiaarm_syscon_base;
 
 static inline uint32_t axxia_mmio_read_32(uintptr_t addr)
 {
@@ -110,10 +109,10 @@ static inline uint32_t axxia_mmio_read_32(uintptr_t addr)
 int
 axxia_is_x9(void)
 {
-	unsigned int pfuse;
+	if (of_find_compatible_node(NULL, NULL, "lsi,axm5616"))
+		return 1;
 
-	pfuse = axxia_mmio_read_32((uintptr_t)(axxiaarm_syscon_base + 0x34));
-	return (0xb == (pfuse & 0x1f));
+	return 0;
 }
 
 struct axxia_pcie {
@@ -204,13 +203,17 @@ static int axxia_pcie_wr_own_conf(struct pcie_port *pp, int where,
 
 static void axxia_pcie_prog_viewport_cfg0(struct pcie_port *pp, u32 busdev)
 {
+	u32 upper_base;
+
 	/* Program viewport 0 : OUTBOUND : CFG0 */
 	axxia_pcie_writel_rc(pp,
 		PCIE_ATU_REGION_OUTBOUND | PCIE_ATU_REGION_INDEX0,
 		PCIE_ATU_VIEWPORT);
 	axxia_pcie_writel_rc(pp, pp->cfg0_base, PCIE_ATU_LOWER_BASE);
-if (!axxia_is_x9())
-	axxia_pcie_writel_rc(pp, (pp->cfg0_base >> 32), PCIE_ATU_UPPER_BASE);
+	/* set upper base bits [1:0] for X9, bits[7:0] for XLF */
+	upper_base = (pp->cfg0_base >> 32);
+	upper_base &= (axxia_is_x9()) ? 0x3 : 0xff;
+	axxia_pcie_writel_rc(pp, upper_base, PCIE_ATU_UPPER_BASE);
 	axxia_pcie_writel_rc(pp, pp->cfg0_base + pp->cfg0_size - 1,
 		PCIE_ATU_LIMIT);
 	axxia_pcie_writel_rc(pp, busdev, PCIE_ATU_LOWER_TARGET);
@@ -222,14 +225,18 @@ if (!axxia_is_x9())
 
 static void axxia_pcie_prog_viewport_cfg1(struct pcie_port *pp, u32 busdev)
 {
+	u32 upper_base;
+
 	/* Program viewport 1 : OUTBOUND : CFG1 */
 	axxia_pcie_writel_rc(pp,
 		PCIE_ATU_REGION_OUTBOUND | PCIE_ATU_REGION_INDEX1,
 		PCIE_ATU_VIEWPORT);
 	axxia_pcie_writel_rc(pp, PCIE_ATU_TYPE_CFG1, PCIE_ATU_CR1);
 	axxia_pcie_writel_rc(pp, pp->cfg1_base, PCIE_ATU_LOWER_BASE);
-if (!axxia_is_x9())
-	axxia_pcie_writel_rc(pp, (pp->cfg1_base >> 32), PCIE_ATU_UPPER_BASE);
+	upper_base = (pp->cfg1_base >> 32);
+	upper_base &= (axxia_is_x9()) ? 0x3 : 0xff;
+	axxia_pcie_writel_rc(pp, upper_base, PCIE_ATU_UPPER_BASE);
+
 	axxia_pcie_writel_rc(pp, pp->cfg1_base + pp->cfg1_size - 1,
 		PCIE_ATU_LIMIT);
 	axxia_pcie_writel_rc(pp, busdev, PCIE_ATU_LOWER_TARGET);
@@ -239,14 +246,17 @@ if (!axxia_is_x9())
 
 static void axxia_pcie_prog_viewport_mem_outbound(struct pcie_port *pp)
 {
+	u32 upper_base;
+
 	/* Program viewport 0 : OUTBOUND : MEM */
 	axxia_pcie_writel_rc(pp,
 		PCIE_ATU_REGION_OUTBOUND | PCIE_ATU_REGION_INDEX2,
 		PCIE_ATU_VIEWPORT);
 	axxia_pcie_writel_rc(pp, PCIE_ATU_TYPE_MEM, PCIE_ATU_CR1);
 	axxia_pcie_writel_rc(pp, pp->mem_mod_base, PCIE_ATU_LOWER_BASE);
-if (!axxia_is_x9())
-	axxia_pcie_writel_rc(pp, (pp->mem_mod_base >> 32), PCIE_ATU_UPPER_BASE);
+	upper_base = (pp->mem_mod_base >> 32);
+	upper_base &= (axxia_is_x9()) ? 0x3 : 0xff;
+	axxia_pcie_writel_rc(pp, upper_base, PCIE_ATU_UPPER_BASE);
 	axxia_pcie_writel_rc(pp, pp->mem_mod_base + pp->mem_size - 1,
 		PCIE_ATU_LIMIT);
 	axxia_pcie_writel_rc(pp, pp->mem_bus_addr, PCIE_ATU_LOWER_TARGET);
@@ -489,7 +499,7 @@ int axxia_pcie_link_up(struct pcie_port *pp)
 
 	axxia_cc_gpreg_readl(pp, PEI_SII_PWR_MGMT_REG, &smlh_state);
 	smlh_state = (smlh_state & PEI_SMLH_LINK_STATE) >> 4;
-	if (smlh_state != 0x11) {
+	if ((smlh_state != 0x11) && (smlh_state != 0x23)) {
 		pr_info("smlh_state = 0x%x\n", smlh_state);
 		pr_err("PCIe LINK IS NOT UP\n");
 		return 0;
@@ -554,11 +564,14 @@ void axxia_pcie_setup_rc(struct pcie_port *pp)
 	val |= PCI_COMMAND_IO | PCI_COMMAND_MEMORY |
 		PCI_COMMAND_MASTER | PCI_COMMAND_SERR;
 	axxia_pcie_writel_rc(pp, val, PCI_COMMAND);
+	axxia_pcie_writel_rc(pp, 0x1037201, 0x8a8);
 
 	/* LTSSM enable */
 	axxia_cc_gpreg_readl(pp, PEI_GENERAL_CORE_CTL_REG, &val);
+	msleep(100);
 	val |= 0x1;
 	axxia_cc_gpreg_writel(pp, 0x1, PEI_GENERAL_CORE_CTL_REG);
+	msleep(100);
 }
 
 static int axxia_pcie_establish_link(struct pcie_port *pp)
@@ -740,11 +753,6 @@ int __init axxia_pcie_host_init(struct pcie_port *pp)
 	LIST_HEAD(res);
 	int i;
 
-	axxiaarm_syscon_base = ioremap(AXXIA_SYSCON_BASE, 0x1024);
-	if (axxiaarm_syscon_base == NULL) {
-		dev_err(pp->dev, "ioremap axxiaarm_syscon_base error!\n");
-		return -EINVAL;
-	}
 	/* Find the address cell size and the number of cells in order to get
 	 * the untranslated address.
 	 */
-- 
1.7.5.4

